### PR TITLE
[GPU] Remote context reuse and internal config update

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**
### Details:
 - Introduced single config class (`ExecutionConfig`) which replaced program build_options, engine_configuration and Config. New config is based on ov::AnyMap and mimics public API.
 - weak_ptr to RemoteContext in RemoteBlob objects is replaced with shared PTR to ensure that context is always alive when blob exits
 - New c-tor for remote context with cldnn::device handle to avoid multiple calls of device query.
 - device name field of remove context is initialized once on creation
 - Custom layers config is loaded in Program instead of engine (including global config)
 - rebote blobs/surfaces cache is now implemented via LRUCache and stores cldnn::memory instead of RemoteBlob to avoid circular shared ptr dependency.
 - Task executor is not a part of engine anymore. It's created for each program individually.

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/remote_blob.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    , m_mem(mem)
+    , m_surf(surf)
+    , m_plane(plane)
+    , m_layout(layout)
+    , m_mem_type(mem_type)
+    , m_memory_object(nullptr)
+    , lockedCounter(0)
+    , lockedHolder(nullptr)
+    , _handle(nullptr) {
+    m_hash = cldnn::hash_combine(0, m_mem);
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/remote_blob.cpp]: Updated

File: src/plugins/intel_gpu/src/plugin/remote_context.cpp [hunk 2]
```diff
@@ -370,26 +100,192 @@ AnyMap ExecutionContextImpl::getParams() const {
...
+    } else {
+        // user will supply shared object handle
+        std::string mem_type = extract_object<std::string>(params, GPU_PARAM_KEY(SHARED_MEM_TYPE));
+
+        bool is_usm = mem_type == GPU_PARAM_VALUE(USM_HOST_BUFFER) ||
+                      mem_type == GPU_PARAM_VALUE(USM_DEVICE_BUFFER) ||
+                      mem_type == GPU_PARAM_VALUE(USM_USER_BUFFER);
+
+        if (is_usm && !m_engine->use_unified_shared_memory()) {
+            IE_THROW(NotAllocated) << "Can't create USM tensor as USM is not supported (or manually disabled) on current device";
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/remote_context.cpp]: Updated

File: src/plugins/intel_gpu/src/plugin/remote_context.cpp [hunk 3]
```diff
@@ -370,26 +100,192 @@ AnyMap ExecutionContextImpl::getParams() const {
...
+                blob_type = BlobType::BT_IMG_SHARED;
+                mem = extract_object<cldnn::shared_handle>(params, GPU_PARAM_KEY(MEM_HANDLE));
+#ifdef _WIN32
+            } else if (GPU_PARAM_VALUE(DX_BUFFER) == mem_type) {
+                blob_type = BlobType::BT_DX_BUF_SHARED;
+                mem = extract_object<cldnn::shared_handle>(params, GPU_PARAM_KEY(DEV_OBJECT_HANDLE));
+                check_if_shared();
+#endif
+            } else {
+                IE_THROW() << "Unsupported shared object type " << mem_type;
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/remote_context.cpp]: Updated

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.hpp [hunk 4]
```diff
@@ -40,27 +40,27 @@ class ocl_engine : public engine {
...
+    static std::shared_ptr<cldnn::engine> create(const device::ptr device, runtime_types runtime_type);
 
 private:
     std::string _extensions;
-    std::unique_ptr<stream> _program_stream;
+    std::unique_ptr<stream> _service_stream;
     std::unique_ptr<cl::UsmHelper> _usm_helper;
 
 #ifdef ENABLE_ONEDNN_FOR_GPU
     mutable std::mutex onednn_mutex;
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.hpp]: Updated

File: src/plugins/intel_gpu/src/plugin/legacy_api_helper.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    } else if (legacy_name == InferenceEngine::GPUConfigParams::KEY_GPU_HOST_TASK_PRIORITY) {
+        ov::Any converted_val{nullptr};
+        auto legacy_val = legacy_property.second.as<std::string>();
+        if (legacy_val == InferenceEngine::GPUConfigParams::GPU_HOST_TASK_PRIORITY_HIGH) {
+            converted_val = ov::hint::Priority::HIGH;
+        } else if (legacy_val == InferenceEngine::GPUConfigParams::GPU_HOST_TASK_PRIORITY_MEDIUM) {
+            converted_val = ov::hint::Priority::MEDIUM;
+        } else if (legacy_val == InferenceEngine::GPUConfigParams::GPU_HOST_TASK_PRIORITY_LOW) {
+            converted_val = ov::hint::Priority::LOW;
+        }
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/legacy_api_helper.cpp:98]: Updated

File: src/plugins/intel_gpu/src/plugin/remote_allocators.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <memory>
+#include "intel_gpu/plugin/remote_context.hpp"
+#include "intel_gpu/plugin/remote_allocators.hpp"
+#include "intel_gpu/plugin/remote_blob.hpp"
+#include "intel_gpu/plugin/itt.hpp"
+#include "intel_gpu/plugin/plugin.hpp"
+#include "intel_gpu/runtime/device_query.hpp"
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/remote_allocators.cpp]: Copy-paste artifact. Removed

File: src/plugins/intel_gpu/src/plugin/remote_allocators.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    }
+};
+
+void USMHostAllocator::unlock(void* handle) noexcept {}
+
+void* USMHostAllocator::alloc(size_t size) noexcept {
+    try {
+        auto td = InferenceEngine::TensorDesc(InferenceEngine::Precision::U8, InferenceEngine::SizeVector{size}, InferenceEngine::Layout::C);
+        InferenceEngine::ParamMap params = {{GPU_PARAM_KEY(SHARED_MEM_TYPE), GPU_PARAM_VALUE(USM_HOST_BUFFER)}};
+        _usm_host_blob = std::dynamic_pointer_cast<InferenceEngine::gpu::USMBlob>(_context->CreateBlob(td, params));
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/remote_allocators.cpp]: Updated

File: src/plugins/intel_gpu/include/intel_gpu/plugin/device_config.hpp [hunk 8]
```diff
@@ -1,105 +0,0 @@
...
-                                                    std::max(1, static_cast<int>(std::thread::hardware_concurrency())), // # of streams
-                                                    1,                                                                  // # of threads per streams
-                                                    InferenceEngine::IStreamsExecutor::ThreadBindingType::HYBRID_AWARE, // thread binding type
-                                                    1,                                                                  // thread binding step
-                                                    0,                                                                  // thread binding offset
-                                                    1,                                                                  // # of threads
-                                                    InferenceEngine::IStreamsExecutor::Config::ANY}),                   // preferred core type
-                                          enable_loop_unrolling(true) {
-        GPU_DEBUG_GET_INSTANCE(debug_config);
-        GPU_DEBUG_IF(debug_config->serialize_compile == 1) {
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/device_config.hpp:46]: Fixed

File: src/plugins/intel_gpu/src/plugin/remote_context.cpp [hunk 9]
```diff
@@ -370,26 +100,191 @@ AnyMap ExecutionContextImpl::getParams() const {
...
-        auto all_devices = device_query.get_available_devices();
-        auto current_device = m_engine->get_device();
-
-        for (auto& kv : all_devices) {
-            if (current_device->is_same(kv.second))
-                return devName + "." + kv.first;
+// For external contexts we try to match underlying handles with default contexts created by plugin to find device name
+std::string RemoteContextImpl::get_device_name(const std::vector<RemoteContextImpl::Ptr>& known_contexts,
+                                               const cldnn::device::ptr current_device) {
+    std::string device_name = "GPU";
```
Comment (ilya-lavrenov) [inline src/plugins/intel_gpu/src/plugin/remote_context.cpp:106]: why do we have a hardcoded `GPU` here? What if plugin name is set differently by user or proxy ?

**Pull request merged**